File: template-assets/javascript/ds_utils.js
class Graph {
    /**
     * A node in a graph.
     * @param {any} val - The value of the node.
     * @param {Graph[]} neighbors - List of neighbor nodes.
     */
    constructor(val = null, neighbors = []) {
        this.val = val;
        this.neighbors = neighbors;
    }

    toString() {
        return `Graph(val=${this.val})`;
    }

    hashCode() {
        const hash = (str) => {
            let hash = 0;
            for (let i = 0; i < str.length; i++) {
                const char = str.charCodeAt(i);
                hash = (hash << 5) - hash + char;
                hash |= 0; // Convert to 32bit integer
            }
            return hash;
        };
        const nodeHashes = Array.from(this.neighbors, (neighbor) => neighbor.hashCode()).sort();
        return hash(`${this.val}:${nodeHashes.join(',')}`);
    }
}

class TreeNode {
    /**
     * A node in a binary tree.
     * @param {any} val - The value of the node.
     * @param {TreeNode|null} left - Left child.
     * @param {TreeNode|null} right - Right child.
     */
    constructor(val = null, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    toString() {
        return `TreeNode(val=${this.val})`;
    }

    hashCode() {
        const hash = (str) => {
            let hash = 0;
            for (let i = 0; i < str.length; i++) {
                const char = str.charCodeAt(i);
                hash = (hash << 5) - hash + char;
                hash |= 0; // Convert to 32bit integer
            }
            return hash;
        };
        const leftHash = this.left ? this.left.hashCode() : 0;
        const rightHash = this.right ? this.right.hashCode() : 0;
        return hash(`${this.val}:${leftHash}:${rightHash}`);
    }
}

class ListNode {
    /**
     * A node in a singly linked list.
     * @param {any} val - The value of the node.
     * @param {ListNode|null} next - The next node.
     */
    constructor(val = null, next = null) {
        this.val = val;
        this.next = next;
    }

    toString() {
        return `ListNode(val=${this.val})`;
    }

    hashCode() {
        const hash = (str) => {
            let hash = 0;
            for (let i = 0; i < str.length; i++) {
                const char = str.charCodeAt(i);
                hash = (hash << 5) - hash + char;
                hash |= 0; // Convert to 32bit integer
            }
            return hash;
        };
        let current = this;
        let hashValue = 0;
        while (current) {
            hashValue = hash(`${hashValue}:${current.val}`);
            current = current.next;
        }
        return hashValue;
    }
}

// Generate a binary tree from a list of values
function generateTree(values) {
    if (!values || values.length === 0) return null;

    const root = new TreeNode(values[0]);
    const queue = [root];
    let i = 1;

    while (queue.length > 0 && i < values.length) {
        const current = queue.shift();

        if (values[i] !== null) {
            current.left = new TreeNode(values[i]);
            queue.push(current.left);
        }
        i++;

        if (i < values.length && values[i] !== null) {
            current.right = new TreeNode(values[i]);
            queue.push(current.right);
        }
        i++;
    }

    return root;
}

// Export a binary tree to a list of values
function exportTree(root) {
    if (!root) return [];

    const result = [];
    const queue = [root];

    while (queue.length > 0) {
        const current = queue.shift();
        if (current) {
            result.push(current.val);
            queue.push(current.left);
            queue.push(current.right);
        } else {
            result.push(null);
        }
    }

    // Remove trailing null values
    while (result[result.length - 1] === null) {
        result.pop();
    }

    return result;
}

// Generate a singly linked list from a list of values
function generateLinkedList(values) {
    if (!values || values.length === 0) return null;

    const head = new ListNode(values[0]);
    let current = head;

    for (let i = 1; i < values.length; i++) {
        current.next = new ListNode(values[i]);
        current = current.next;
    }

    return head;
}

// Export a singly linked list to a list of values
function exportLinkedList(head) {
    const result = [];
    let current = head;

    while (current) {
        result.push(current.val);
        current = current.next;
    }

    return result;
}

// Generate a graph from a list of edges
function generateGraph(edges, directed = false) {
    const nodes = new Map();

    edges.forEach(([u, v]) => {
        if (!nodes.has(u)) nodes.set(u, new Graph(u));
        if (!nodes.has(v)) nodes.set(v, new Graph(v));

        nodes.get(u).neighbors.push(nodes.get(v));
        if (!directed) {
            nodes.get(v).neighbors.push(nodes.get(u));
        }
    });

    return nodes;
}

// Export a graph to a list of edges
function exportGraph(nodes, directed = false) {
    const edges = [];
    const visited = new Set();

    nodes.forEach((node) => {
        node.neighbors.forEach((neighbor) => {
            if (!visited.has(`${neighbor.val}-${node.val}`)) {
                edges.push([node.val, neighbor.val]);
                if (!directed) {
                    visited.add(`${node.val}-${neighbor.val}`);
                }
            }
        });
    });

    return edges;
}

// Exports
module.exports = {
    Graph,
    TreeNode,
    ListNode,
    generateTree,
    exportTree,
    generateLinkedList,
    exportLinkedList,
    generateGraph,
    exportGraph,
};

File: template-assets/python/ds_utils.py
from typing import Dict, TypeVar, Generic, List, Optional, Union

T = TypeVar("T")  # Generic type for the value of the node

class TreeNode(Generic[T]):
    def __init__(self, val: T = None, left: Optional['TreeNode[T]'] = None, right: Optional['TreeNode[T]'] = None):
        self.val = val
        self.left = left
        self.right = right

    def __eq__(self, other):
        if not isinstance(other, TreeNode):
            return False
        return (
            self.val == other.val
            and self.left == other.left
            and self.right == other.right
        )

    def __repr__(self):
        return f"TreeNode(val={self.val}, left={self.left}, right={self.right})"

    def __hash__(self):
        return hash((self.val, self.left, self.right))

class ListNode(Generic[T]):
    def __init__(self, val: T = None, next: Optional['ListNode[T]'] = None):
        self.val = val
        self.next = next

    def __eq__(self, other):
        if not isinstance(other, ListNode):
            return False
        current_self, current_other = self, other
        while current_self and current_other:
            if current_self.val != current_other.val:
                return False
            current_self = current_self.next
            current_other = current_other.next
        return current_self is None and current_other is None

    def __repr__(self):
        return f"ListNode(val={self.val}, next={self.next})"

    def __hash__(self):
        current = self
        hash_value = 0
        while current:
            hash_value = hash((hash_value, current.val))
            current = current.next
        return hash_value

class Graph(Generic[T]):
    def __init__(self):
        self.adj_list: Dict[T, List[Union[T, 'Graph[T]', 'TreeNode[T]', 'ListNode[T]']]] = {}

    def add_edge(self, u: T, v: Union[T, 'Graph[T]', 'TreeNode[T]', 'ListNode[T]']):
        if u not in self.adj_list:
            self.adj_list[u] = []
        self.adj_list[u].append(v)

    def __eq__(self, other):
        if not isinstance(other, Graph):
            return False
        if set(self.adj_list.keys()) != set(other.adj_list.keys()):
            return False
        for key in self.adj_list:
            if sorted(self.adj_list[key], key=str) != sorted(other.adj_list[key], key=str):
                return False
        return True

    def __repr__(self):
        return f"Graph(adj_list={self.adj_list})"

    def __hash__(self):
        return hash(frozenset((u, frozenset(neighbors)) for u, neighbors in self.adj_list.items()))

# Generate a tree that supports nested data structures
def generate_tree(values: List[Optional[Union[T, dict]]]) -> Optional[TreeNode[T]]:
    """
    Generate a binary tree from a list of values or nested structures.

    :param values: List of values or dictionaries (e.g., [1, {"val": 2, "left": None}, 3]).
    :return: Root of the constructed TreeNode.
    """
    if not values:
        return None

    root = TreeNode(values[0] if not isinstance(values[0], dict) else TreeNode(**values[0]))
    queue = [root]
    i = 1

    while queue and i < len(values):
        current = queue.pop(0)

        if i < len(values) and values[i] is not None:
            current.left = TreeNode(values[i] if not isinstance(values[i], dict) else TreeNode(**values[i]))
            queue.append(current.left)
        i += 1

        if i < len(values) and values[i] is not None:
            current.right = TreeNode(values[i] if not isinstance(values[i], dict) else TreeNode(**values[i]))
            queue.append(current.right)
        i += 1

    return root

# Generate a graph that supports nested data structures
def generate_graph(edges: List[List[Union[T, dict]]]) -> Graph[T]:
    """
    Generate a graph from a list of directed edges with nested structures.

    :param edges: List of edges (e.g., [[1, 2], [3, {"val": 4}]]).
    :return: A Graph object represented as an adjacency list.
    """
    graph = Graph[T]()
    for edge in edges:
        if len(edge) != 2:
            raise ValueError(f"Invalid edge format: {edge}. Each edge must have exactly two elements.")
        u = edge[0] if not isinstance(edge[0], dict) else TreeNode(**edge[0])
        v = edge[1] if not isinstance(edge[1], dict) else TreeNode(**edge[1])
        graph.add_edge(u, v)
    return graph

# Export tree structure back to a list
def export_tree(root: Optional[TreeNode[T]]) -> List[Optional[Union[T, dict]]]:
    """
    Export a binary tree to a list of values or nested structures.

    :param root: Root of the TreeNode.
    :return: List of values or nested structures.
    """
    if not root:
        return []

    result = []
    queue = [root]

    while queue:
        current = queue.pop(0)
        if current:
            result.append(current.val if not isinstance(current.val, TreeNode) else export_tree(current.val))
            queue.append(current.left)
            queue.append(current.right)
        else:
            result.append(None)

    # Remove trailing None values for cleaner output
    while result and result[-1] is None:
        result.pop()

    return result

# Export graph back to edges
def export_graph(graph: Graph[T]) -> List[List[Union[T, dict]]]:
    """
    Export a graph to a list of directed edges with nested structures.

    :param graph: A Graph object represented as an adjacency list.
    :return: List of edges with nested structures.
    """
    edges = []
    for u, neighbors in graph.adj_list.items():
        for v in neighbors:
            edges.append([u, v])
    return edges

# Generate a linked list from a list of values
def generate_linked_list(values: List[Optional[T]]) -> Optional[ListNode[T]]:
    """
    Generate a singly linked list from a list of values.

    :param values: List of values (e.g., [1, 2, 3]).
    :return: Head of the constructed ListNode.
    """
    if not values:
        return None

    head = ListNode(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next

    return head

# Export a linked list to a list of values
def export_linked_list(head: Optional[ListNode[T]]) -> List[Optional[T]]:
    """
    Export a singly linked list to a list of values.

    :param head: Head of the ListNode.
    :return: List of values.
    """
    result = []
    current = head
    while current:
        result.append(current.val)
        current = current.next

    return result
